src/main/java/io/vertx/core/http/impl/pool/Pool.java

```diff
@@ -0,0 +1,341 @@
+      if (waiter == null) {
+        break;
+      }
+      if (metric != null) {
+        metrics.dequeueRequest(metric, waiter.metric);
+      }
+      if (capacity > 0) {
+        capacity--;
+        Holder<C> conn = available.peek();
+        if (--conn.capacity == 0) {
+          available.poll();
+        }
+        waiters.poll();
+        ContextImpl ctx = conn.context;
+        ctx.nettyEventLoop().execute(() -> {
+          if (connector.isValid(conn.connection)) {
+            boolean handled = deliverToWaiter(conn, waiter);
+            if (!handled) {
+              synchronized (Pool.this) {
+                recycleConnection(conn, 1,false);
+                checkPending();
+              }
+            }
+          } else {
+            synchronized (Pool.this) {
+              waiters.add(waiter);
+              closed(conn);
+            }
+          }
+        });
+      } else if (weight < maxWeight) {
+        waiters.poll();
+        weight += createConnection(waiter);
+      } else {
+        break;
+      }
+    }
+  }
+
+  private long createConnection(Waiter<C> waiter) {
+    Holder<C> holder  = new Holder<>();
+    ConnectionListener<C> listener = new ConnectionListener<C>() {
+      @Override
+      public void onConnectSuccess(C conn, long concurrency, Channel channel, ContextImpl context, long initialWeight, long actualWeight) {
+        waiter.initConnection(context, conn);
+        boolean usable;
+        synchronized (Pool.this) {
+          usable = initConnection(waiter, holder, context, concurrency, conn, channel, initialWeight, actualWeight);
+        }
+        if (usable) {
+          boolean consumed = deliverToWaiter(holder, waiter);
+          synchronized (Pool.this) {
+            if (!consumed) {
+              synchronized (this) {
+                recycleConnection(holder, 1,false);
+              }
+            }
+            checkPending();
+          }
+        } else {
+          synchronized (Pool.this) {
+            checkPending();
+          }
+        }
+      }
+      @Override
+      public void onConnectFailure(ContextImpl context, Throwable err, long weight) {
+        waiter.handleFailure(context, err);
+        synchronized (Pool.this) {
+          Pool.this.weight -= weight;
+          all.remove(holder);
+          checkPending();
+          checkClose();
+        }
+      }
+      @Override
+      public void onConcurrencyChange(long concurrency) {
+        synchronized (Pool.this) {
+          if (holder.concurrency < concurrency) {
+            long diff = concurrency - holder.concurrency;
+            capacity += diff;
+            if (holder.capacity == 0) {
+              available.add(holder);
+            }
+            holder.capacity += diff;
+            holder.concurrency = concurrency;
+            checkPending();
+          } else if (holder.concurrency > concurrency) {
+            throw new UnsupportedOperationException("Not yet implemented");
```

Comment:
note, this is a case when an HTTP/2 connection send a max concurrent stream = 20 and then downgrade to a lower version, it should be supported at some point but it seems difficult and I don't believe it is an actual case

Reply:
What would the Vert.x HTTP server do?

Reply:
what do you mean ?

Reply:
You said 

> In practice the server will rather send a GO_AWAY frame.

 Is that what a Vert.x HTTP/2 server would do?